File: bun-baseline/order-store.ts
/**
 * In-memory order state management
 * Stores order status and provides thread-safe access
 */

export type OrderStatus = "received" | "processing" | "completed" | "failed";

export interface Order {
  id: string;
  status: OrderStatus;
  createdAt: Date;
  updatedAt: Date;
  retryCount: number;
  data?: any;
}

class OrderStore {
  private orders: Map<string, Order> = new Map();

  create(id: string, data?: any): Order {
    const order: Order = {
      id,
      status: "received",
      createdAt: new Date(),
      updatedAt: new Date(),
      retryCount: 0,
      data,
    };
    this.orders.set(id, order);
    return order;
  }

  get(id: string): Order | undefined {
    return this.orders.get(id);
  }

  updateStatus(id: string, status: OrderStatus): void {
    const order = this.orders.get(id);
    if (order) {
      order.status = status;
      order.updatedAt = new Date();
      this.orders.set(id, order);
    }
  }

  incrementRetry(id: string): void {
    const order = this.orders.get(id);
    if (order) {
      order.retryCount++;
      this.orders.set(id, order);
    }
  }

  getAll(): Order[] {
    return Array.from(this.orders.values());
  }

  clear(): void {
    this.orders.clear();
  }
}

export const orderStore = new OrderStore();

File: elixir-candidate/lib/elixir_candidate/order_store.ex
defmodule ElixirCandidate.OrderStore do
  @moduledoc """
  ETS-based order state management
  Each order is stored with its status and metadata
  """
  use GenServer

  @table_name :orders

  # Client API

  def start_link(_opts) do
    GenServer.start_link(__MODULE__, :ok, name: __MODULE__)
  end

  def create(order_id, data \\ %{}) do
    order = %{
      id: order_id,
      status: :received,
      created_at: DateTime.utc_now(),
      updated_at: DateTime.utc_now(),
      retry_count: 0,
      data: data
    }
    :ets.insert(@table_name, {order_id, order})
    order
  end

  def get(order_id) do
    case :ets.lookup(@table_name, order_id) do
      [{^order_id, order}] -> {:ok, order}
      [] -> {:error, :not_found}
    end
  end

  def update_status(order_id, status) do
    case get(order_id) do
      {:ok, order} ->
        updated_order = %{order | status: status, updated_at: DateTime.utc_now()}
        :ets.insert(@table_name, {order_id, updated_order})
        {:ok, updated_order}
      error -> error
    end
  end

  def increment_retry(order_id) do
    case get(order_id) do
      {:ok, order} ->
        updated_order = %{order | retry_count: order.retry_count + 1}
        :ets.insert(@table_name, {order_id, updated_order})
        {:ok, updated_order}
      error -> error
    end
  end

  def get_all do
    :ets.tab2list(@table_name)
    |> Enum.map(fn {_id, order} -> order end)
  end

  def stats do
    orders = get_all()
    %{
      total: length(orders),
      received: Enum.count(orders, &(&1.status == :received)),
      processing: Enum.count(orders, &(&1.status == :processing)),
      completed: Enum.count(orders, &(&1.status == :completed)),
      failed: Enum.count(orders, &(&1.status == :failed))
    }
  end

  # Server Callbacks

  @impl true
  def init(:ok) do
    :ets.new(@table_name, [:named_table, :public, read_concurrency: true])
    {:ok, %{}}
  end
end
